### PR TITLE
feat/multiple-conditions-service

### LINKED ISSUE #1917: Multiple condition on flow trigger does not handle consistently the flow state
### Expected Behavior

I have a flow that need to have two flows dependencies in success, to run.
More precisely: 
- Given flow A and B in success state during the last time window, flow C will be trigger and run
- Given flow A in error and flow B in success state during the last time window, flow C will **not** be trigger and will not run
- Given flow A and flow B in error state during the last minute, flow C will **not** be trigger and will not run

### Actual Behaviour

When flow A is in error state and flow B is in success state, flow C will run when it should not.

### Steps To Reproduce

In the same time window configured in flow C:
1. Run flow A that will fail
2. Run flow B that will succeed
3. Check that flow C has been executed when it should not

### Environment Information

- Kestra Version: 0.9.8

dep graph:
image

### Example flow

Time window for testing: 1 minute
```yaml
id: trigger_test
namespace: my.namespace

tasks:
  - id: test
    type: io.kestra.core.tasks.log.Log
    message: "OK"

triggers:
  - id: multiple-listen-flow
    type: io.kestra.core.models.triggers.types.Flow
    conditions:
      - type: io.kestra.core.models.conditions.types.ExecutionStatusCondition
        in:
          - SUCCESS
      - id: multiple
        type: io.kestra.core.models.conditions.types.MultipleCondition
        window: PT1M
        windowAdvance: PT0S
        conditions:
          flow_a:
            type: io.kestra.core.models.conditions.types.ExecutionFlowCondition
            namespace: my.namespace
            flowId: flow_a
          flow_b:
            type: io.kestra.core.models.conditions.types.ExecutionFlowCondition
            namespace: my.namespace
            flowId: flow_b

```

flow A:
```yaml
id: flow_a
namespace: my.namespace

tasks:
  - id: test
    type: io.kestra.core.tasks.log.Log
    message: "OK"
  - id: fail
    type: io.kestra.core.tasks.executions.Fail
```

flow B:
```yaml
id: flow_b
namespace: my.namespace

tasks:
  - id: test
    type: io.kestra.core.tasks.log.Log
    message: "OK"
  #- id: fail
  #  type: io.kestra.core.tasks.executions.Fail
```

### PR DESCRIPTION
**Pull request by brian-mulier-p**
closes #1917

### CODE REVIEW

File: core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.ToString;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractFlowTriggerService<S extends MultipleConditionStorageInterface> {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java:20]: I don't see the point of having a type parameter here, we usually inject the interface and depending on the runner and implementation is choosed.

File: core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Inject
+    private FlowService flowService;
+
+    public AbstractFlowTriggerService(Optional<S> multipleConditionStorage, ConditionService conditionService, RunContextFactory runContextFactory, FlowService flowService) {
+        this.multipleConditionStorage = multipleConditionStorage;
+        this.conditionService = conditionService;
+        this.runContextFactory = runContextFactory;
+        this.flowService = flowService;
+    }
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java:38]: Please choose between field injection or constructor injection but don't use both.

File: core/src/main/java/io/kestra/core/services/FlowService.java [hunk 3]
```diff
@@ -307,28 +184,11 @@ private static Object fixSnakeYaml(Object object) {
...
-        private final AbstractTrigger trigger;
-    }
-
     @AllArgsConstructor
     @Getter
     @ToString
     public static class FlowWithFlowTrigger {
         private final Flow flow;
         private final io.kestra.core.models.triggers.types.Flow trigger;
     }
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/services/FlowService.java:193]: Shouldn't this on be moved to the new AbstractFlowTriggerService?

File: core/src/main/java/io/kestra/core/services/FlowService.java [hunk 4]
```diff
@@ -83,131 +80,11 @@ private Stream<Flow> keepLastVersionCollector(Stream<Flow> stream) {
...
-                    (io.kestra.core.models.triggers.types.Flow) f.getTrigger()
-                )
-            )
-            .collect(Collectors.toList());
-    }
-
     protected boolean removeUnwanted(Flow f, Execution execution) {
         // we don't allow recursive
         return !f.uidWithoutRevision().equals(Flow.uidWithoutRevision(execution));
     }
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/services/FlowService.java:86]: I think this method is only used on the new AbstractFlowTriggerService, if so, move it there

File: core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+import lombok.ToString;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractFlowTriggerService<S extends MultipleConditionStorageInterface> {
+    @Inject
+    protected Optional<S> multipleConditionStorage;
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java:22]: It seems that all runner have a multiple condition storage so why is it optional?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Also please remove the changes to package-lock.json as they are unrelated

File: core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java (hunk 1 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java:20]: That's something I forgot to remove as before the save & delete methods weren't in the interface so I had to have access to the impl directly to be able to do such action and that was preventing the need to cast

File: core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java (hunk 2 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/services/AbstractFlowTriggerService.java:38]: Then I'll have to stick with constructor as Kafka provides its multipleConditionStorage based on a State Store which is a runtime thing :'(

File: jdbc/src/main/java/io/kestra/jdbc/runner/JdbcFlowTriggerService.java [hunk 6]
```diff
@@ -1,17 +1,9 @@
...
 
-import java.util.Optional;
-
 @JdbcRunnerEnabled
 @Singleton
-public class JdbcFlowTriggerService extends AbstractFlowTriggerService<AbstractJdbcMultipleConditionStorage> {
-    public JdbcFlowTriggerService(Optional<AbstractJdbcMultipleConditionStorage> multipleConditionStorage, ConditionService conditionService, RunContextFactory runContextFactory, FlowService flowService) {
-        super(multipleConditionStorage, conditionService, runContextFactory, flowService);
-    }
+public class JdbcFlowTriggerService extends AbstractFlowTriggerService {
```
Comment (loicmathieu) [inline jdbc/src/main/java/io/kestra/jdbc/runner/JdbcFlowTriggerService.java:8]: If both JdbcFlowTriggerService and MemoryFlowTriggerService are the same maybe you can create a DefaultFlowTriggerService (or just a FlowTrigerServiceImpl) that will be the default and replaces inside the Kafka runner.

**Review by loicmathieu — APPROVED** [review]

**Pull request merged**